test/functional/fixtures/regression/gh-4516/testcafe-fixtures/index.js

```diff
@@ -0,0 +1,49 @@
+import { RequestHook, Selector } from 'testcafe';
+
+export default class CustomHook extends RequestHook {
+    constructor (config) {
+        super(null, config);
+
+        this.pendingAjaxRequestIds = new Set();
+        this.hasAjaxRequests       = false;
+    }
+
+    onRequest (event) {
+        if (event.isAjax) {
+            this.pendingAjaxRequestIds.add(event._requestInfo.requestId);
+            this.hasAjaxRequests = true;
+        }
+    }
+
+    onResponse (event) {
+        this.pendingAjaxRequestIds.delete(event.requestId);
+    }
+}
+
+const hook1 = new CustomHook();
+const hook2 = new CustomHook({});
+const hook3 = new CustomHook({ includeHeaders: true });
+
+fixture `GH-4516`
+    .page `http://localhost:3000/fixtures/regression/gh-4516/pages/index.html`;
+
+test.requestHooks(hook1)('Without config', async t => {
+    await t
+        .expect(Selector('#result').visible).ok()
+        .expect(hook1.hasAjaxRequests).ok()
+        .expect(hook1.pendingAjaxRequestIds.size).eql(0);
+});
+
+test.requestHooks(hook2)('With empty config', async t => {
+    await t
+        .expect(Selector('#result').visible).ok()
+        .expect(hook2.hasAjaxRequests).ok()
```

Comment:
To prevent unstable tests return the checked properties via [ReExecutablePromise](https://github.com/DevExpress/testcafe/blob/master/src/utils/re-executable-promise.js).

See the similar behavior here - https://github.com/DevExpress/testcafe/blob/master/test/functional/fixtures/api/es-next/request-hooks/testcafe-fixtures/api/execution-order.js#L25

Reply:
If you see the first commit, to prevent an unstable tests return I've added the `#result` selector visibility check.

Reply:
Yes, I've missed that.  Visibility checking is enough condition for stable tests.